support/k8s/k8s-workload-registrar/mode-crd/controllers/endpoint_controller.go

```diff
@@ -0,0 +1,209 @@
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// EndpointReconcilerConfig holds the config passed in when creating the reconciler
+type EndpointReconcilerConfig struct {
+	Ctx                context.Context
+	DisabledNamespaces []string
+	Log                logrus.FieldLogger
+	Mgr                ctrl.Manager
+	PodLabel           string
+	PodAnnotation      string
+}
+
+// EndpointReconciler holds the runtime configuration and state of this controller
+type EndpointReconciler struct {
+	client.Client
+	c EndpointReconcilerConfig
+}
+
+// NewEndpointReconciler creates a new EndpointReconciler object
+func AddEndpointReconciler(config EndpointReconcilerConfig) error {
+	r := &EndpointReconciler{
+		Client: config.Mgr.GetClient(),
+		c:      config,
+	}
+
+	err := ctrl.NewControllerManagedBy(config.Mgr).
+		For(&corev1.Endpoints{}).
+		Complete(r)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile steps through the endpoints for each service and adds the name of the service as
+// a DNS name to the SPIFFE ID CRD
+func (e *EndpointReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	if containsString(e.c.DisabledNamespaces, req.NamespacedName.Namespace) {
+		return ctrl.Result{}, nil
+	}
+
+	endpoints := corev1.Endpoints{}
+
+	if err := e.Get(e.c.Ctx, req.NamespacedName, &endpoints); err != nil {
+		if errors.IsNotFound(err) {
+			// Delete event
+			if err := e.deleteExternalResources(req.NamespacedName); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+
+		e.c.Log.WithError(err).Error("Unable to fetch Endpoints")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	svcName := getServiceDNSName(req.NamespacedName)
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			if address.TargetRef == nil {
+				continue
+			}
+
+			// Get SPIFFE ID resource associated with this endpoint
+			spiffeIDList := spiffeidv1beta1.SpiffeIDList{}
+			labelSelector := labels.Set(map[string]string{
+				"podUid": string(address.TargetRef.UID),
+			})
+			err := e.List(e.c.Ctx, &spiffeIDList, &client.ListOptions{
+				LabelSelector: labelSelector.AsSelector(),
+			})
+			if err != nil {
+				e.c.Log.WithError(err).Error("Error getting spiffeid list")
+				return ctrl.Result{}, err
+			}
+
+			// If there are no SPIFFE ID resources associated with this endpoint, we may need to requeue
+			// Its possible this reconcile loop ran before the SPIFFE ID resource was generated
+			if len(spiffeIDList.Items) == 0 {
+				return ctrl.Result{
+					Requeue: e.requeue(address.TargetRef.Name, address.TargetRef.Namespace),
+				}, nil
+			}
+
+			// Iterate through the list of SPIFFE ID resources and update to add the DNS name
+			for _, spiffeID := range spiffeIDList.Items {
+				if !containsString(spiffeID.Spec.DnsNames[1:], svcName) {
```

Comment:
why is the first element of DnsNames in the spec not considered when determining if the service name exists in the list?

Reply:
It is statically populated with the name of the pod: https://github.com/faisal-memon/spire/blob/master/support/k8s/k8s-workload-registrar/mode-crd/controllers/pod_controller.go#L121

I can remove the `1:` optimization to make the the code a little safer.

Reply:
Yeah, i think that would be much less surprising.

Reply:
done: https://github.com/spiffe/spire/pull/1581/commits/bd840d62c70f6dc87b4ba8f706c7effa583e75df